src/Workspaces/Core/Portable/FindSymbols/SyntaxTree/SyntaxTreeIndex.cs

```diff
@@ -11,80 +11,93 @@ namespace Microsoft.CodeAnalysis.FindSymbols
             _contextInfo = contextInfo;
             _declarationInfo = declarationInfo;
         }
 
-        /// <summary>
-        /// snapshot based cache to guarantee same info is returned without re-calculating for 
-        /// same solution snapshot.
-        /// 
-        /// since document will be re-created per new solution, this should go away as soon as 
-        /// there is any change on workspace.
-        /// </summary>
-        private static readonly ConditionalWeakTable<Document, SyntaxTreeIndex> s_infoCache
-            = new ConditionalWeakTable<Document, SyntaxTreeIndex>();
+        private static readonly ConditionalWeakTable<Document, SyntaxTreeIndex> s_documentToIndex = new ConditionalWeakTable<Document, SyntaxTreeIndex>();
+        private static readonly ConditionalWeakTable<DocumentId, SyntaxTreeIndex> s_documentIdToIndex = new ConditionalWeakTable<DocumentId, SyntaxTreeIndex>();
 
         public static async Task PrecalculateAsync(Document document, CancellationToken cancellationToken)
         {
             Contract.Requires(document.IsFromPrimaryBranch());
 
-            // we already have information. move on
-            if (await PrecalculatedAsync(document, cancellationToken).ConfigureAwait(false))
+            var checksum = await GetChecksumAsync(document, cancellationToken).ConfigureAwait(false);
+
+            // Check if we've already created and persisted the index for this document.
+            if (await PrecalculatedAsync(document, checksum, cancellationToken).ConfigureAwait(false))
             {
                 return;
             }
 
-            var data = await CreateInfoAsync(document, cancellationToken).ConfigureAwait(false);
+            // If not, create and save the index.
+            var data = await CreateIndexAsync(document, checksum, cancellationToken).ConfigureAwait(false);
             await data.SaveAsync(document, cancellationToken).ConfigureAwait(false);
         }
 
-        private static async Task<SyntaxTreeIndex> GetIndexAsync(
+        public static async Task<SyntaxTreeIndex> GetIndexAsync(
             Document document,
-            ConditionalWeakTable<Document, SyntaxTreeIndex> cache,
-            Func<Document, CancellationToken, Task<SyntaxTreeIndex>> generator,
             CancellationToken cancellationToken)
         {
-            if (cache.TryGetValue(document, out var info))
+            // See if we already cached an index with this direct document index.  If so we can just
+            // return it with no additional work.
+            if (!s_documentToIndex.TryGetValue(document, out var index))
             {
-                return info;
+                index = await GetIndexWorkerAsync(document, cancellationToken).ConfigureAwait(false);
+
+                // Populate our caches with this data.
+                s_documentToIndex.GetValue(document, _ => index);
+                s_documentIdToIndex.GetValue(document.Id, _ => index);
```

Comment:
hmm.. I vaguely remember that one index always in memory since it was so small, and another one saved to disk.. not sure that memory is true or not.